### DISCUSSION TITLE
RFC: allow throwing `PublicError` instances inside server actions

### DISCUSSION THREAD

**Discussion by remorses** [vercel/next.js#55442] [Ideas]
### Goals

Make error handling in server actions simpler by throwing instances of a `PublicError` class (exported from `next`)

Currently error messages get stripped away from server actions errors: https://x.com/asidorenko_/status/1702684047879626792?s=20

### Background

Currently error messages get stripped away from server actions errors: https://x.com/asidorenko_/status/1702684047879626792?s=20

The alternative to this proposal would be to return an object with an error field, but this is not idiomatic in JavaScript and most existing code use exceptions instead

Migrating user code to server actions with this RFC would mean replacing some `throw new Error` with `throw new PublicError`

Without this proposal you would need to wrap all server actions with a `try catch` block and manually handle each error type, which would mean implementing this RFC in user code (or worse, passing all errors down to the client, possibly exposing secret data)

### Proposal

Add a `PublicError` error class (or a different name, see [x poll](https://twitter.com/__morse/status/1702687492372078940)) that retains its error message and properties when thrown inside a server action

The client code can then handle these errors, in case the server action is called directly by using a `catch` block or using the `useFormStatus` hook when using the `action` prop

To make error handling easier i would also suggest turning any non `PublicError` into a `PublicError('Unexpected error')` so that the user will never see cryptic error messages like `The specific error message is omitted ...`

### Example

```tsx
import { PublicError } from 'next'
import toast from 'react-hot-toast'

export default function ServerComponent() {

  async function myAction(userId) {
    'use server'
    const isLoggedIn = await authenticate(userId)
    if (!isLoggedIn) throw new PublicError(`forbidden`)
    // ...
  }
  return <button onClick={() => myAction(userId).catch(e => toast.error(e.message))} />
}
```

**Comment by gerynugrh** [comment]
Perfectly solve a problem that we have in mind, would love to see a way to configure error obfuscation behavior
